cmd/xl-v1-list-objects-heal.go

```diff
@@ -205,3 +206,252 @@ func (xl xlObjects) ListObjectsHeal(bucket, prefix, marker, delimiter string, ma
+	// multipartPrefixPath should have a trailing '/' when prefix = "".
+	if prefix == "" {
+		multipartPrefixPath += slashSeparator
+	}
+
+	multipartMarkerPath := ""
+	if keyMarker != "" {
+		multipartMarkerPath = pathJoin(bucket, keyMarker)
+	}
+
+	// `heal bool` is used to differentiate listing of incomplete
+	// uploads (and parts) from a regular listing of incomplete
+	// parts by client SDKs or mc-like commands, within a treewalk
+	// pool.
+	heal := true
+	var walkerCh chan treeWalkResult
+	var walkerDoneCh chan struct{}
+	// Check if we have room left to send more uploads.
+	if maxUploads > 0 {
+		walkerCh, walkerDoneCh = xl.listPool.Release(listParams{
+			bucket:    minioMetaMultipartBucket,
+			recursive: recursive,
+			marker:    multipartMarkerPath,
+			prefix:    multipartPrefixPath,
+			heal:      heal,
+		})
+		if walkerCh == nil {
+			walkerDoneCh = make(chan struct{})
+			isLeaf := xl.isMultipartUpload
+			listDir := listDirFactory(isLeaf, xlTreeWalkIgnoredErrs,
+				xl.getLoadBalancedDisks()...)
+			walkerCh = startTreeWalk(minioMetaMultipartBucket,
+				multipartPrefixPath, multipartMarkerPath,
+				recursive, listDir, isLeaf, walkerDoneCh)
+		}
+		// Collect uploads until maxUploads limit is reached.
+		for maxUploads > 0 {
+			walkResult, ok := <-walkerCh
+			if !ok {
+				// No more entries, tree walk channel
+				// is closed.
+				eof = true
+				break
+			}
+
+			// For any error during tree walk we should
+			// return right away.
+			if walkResult.err != nil {
+				// Ignore errors like errDiskNotFound
+				// and errFileNotFound.
+				if isErrIgnored(walkResult.err,
+					xlTreeWalkIgnoredErrs...) {
+					continue
+				}
+				return ListMultipartsInfo{}, walkResult.err
+			}
+			entry := strings.TrimPrefix(walkResult.entry,
+				retainSlash(bucket))
+			// Skip entries that are not object directory.
+			if hasSuffix(walkResult.entry, slashSeparator) {
+				uploads = append(uploads, uploadMetadata{
+					Object: entry,
+				})
+				maxUploads--
+				if maxUploads == 0 {
+					eof = true
+					break
+				}
+				continue
+			}
+			var newUploads []uploadMetadata
+			var end bool
+			uploadIDMarker = ""
+
+			// For an object entry we get all its
+			// pending uploadIDs.
+			entryLock := globalNSMutex.NewNSLock(minioMetaMultipartBucket,
+				pathJoin(bucket, entry))
+			entryLock.RLock()
+			var disk StorageAPI
+			for _, disk = range xl.getLoadBalancedDisks() {
+				if disk == nil {
+					continue
+				}
+				newUploads, end, err = listMultipartUploadIDs(bucket, entry,
+					uploadIDMarker, maxUploads, disk)
+				if err == nil {
+					break
+				}
+				if isErrIgnored(err, objMetadataOpIgnoredErrs...) {
+					continue
+				}
+				break
+			}
+			entryLock.RUnlock()
+			if err != nil {
+				if isErrIgnored(err, xlTreeWalkIgnoredErrs...) {
+					continue
+				}
+				return ListMultipartsInfo{}, err
+			}
+			uploads = append(uploads, newUploads...)
+			maxUploads -= len(newUploads)
```

Comment:
Looks like `maxUploads` can go negative here, putting in more entries in `uploads` than was requested.